### PR TITLE
Update of strip overlap check for different product ID

### PR DESCRIPTION
**Pull request by mmasciov**
### PR description:

As follow-up to to [HLT JIRA 3534](https://its.cern.ch/jira/browse/CMSHLT-3534), PR https://github.com/cms-sw/cmssw/pull/48403 (and specifically to https://github.com/cms-sw/cmssw/pull/48403/#discussion_r2255199103, by @slava77), this PR aims at using `sameDetModule` in place of `geographicalId` to better deal with 2D hits, to ensure that the overlap check introduced in PR #48403 is closer to the identity-based method used for strips with the same product ID (impact was estimated ~1e-5 in [HLT JIRA 3534](https://its.cern.ch/jira/browse/CMSHLT-3534)).

#### PR validation:

This was checked with https://github.com/missirol/hltScripts/blob/a89bc26a7d18c6e8f8582878ef489883cf84bd7a/hltTests/testCMSHLT3534.sh provided by @missirol

After the change, the number of hltPFMuonMerging tracks found in this event is indeed identical (111).

A backport to 150X will follow.

FYI: @missirol, @slava77, @cms-sw/tracking-pog-l2

### CODE REVIEW

**Comment by mmasciov** [comment]
please test

**Comment by mmusich** [comment]
type bug-fix

File: DataFormats/TrackerRecHit2D/src/SiStripMatchedRecHit2D.cc [hunk 1]
```diff
@@ -22,7 +22,7 @@ bool SiStripMatchedRecHit2D::sharesInput(const TrackingRecHit* other, SharedInpu
   if (monoClusterRef().id() == otherClus.id() || stereoClusterRef().id() == otherClus.id())
     return (otherClus == stereoClusterRef()) || (otherClus == monoClusterRef());
   else {
-    const bool sameDetId = (geographicalId() == other->geographicalId());
+    const bool sameDetId = sameDetModule(*other);
```
Comment (missirol) [inline DataFormats/TrackerRecHit2D/src/SiStripMatchedRecHit2D.cc]: Earlier in this same function, `sameDetModule` is required to be `true` (otherwise, the function returns `false` right away), so it seems that this part could be simplified ? (`sameDetId` would always be `true`)

File: DataFormats/TrackerRecHit2D/src/VectorHit.cc [hunk 2]
```diff
@@ -78,7 +78,7 @@ bool VectorHit::sharesInput(const TrackingRecHit* other, SharedInputType what) c
   else {
     bool lowerOverlap = false;
     bool upperOverlap = false;
-    if (geographicalId() == other->geographicalId()) {
+    if (sameDetModule(*other)) {
```
Comment (missirol) [inline DataFormats/TrackerRecHit2D/src/VectorHit.cc]: Same as above.

> Earlier in this same function, `sameDetModule` is required to be `true` (otherwise, the function returns `false` right away), so it seems that this part could be simplified ?

Comment (mmasciov) [reply DataFormats/TrackerRecHit2D/src/VectorHit.cc]: Thanks! Done in 18f9ae2230781749a3794fe2730377f4d9e983cb.

File: DataFormats/TrackerRecHit2D/src/SiStripMatchedRecHit2D.cc (hunk 1 above)
Comment (mmasciov) [reply DataFormats/TrackerRecHit2D/src/SiStripMatchedRecHit2D.cc]: Thanks! Done in 18f9ae2230781749a3794fe2730377f4d9e983cb.

**Comment by mmasciov** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by mmasciov** [comment]
please test

**Comment by mmasciov** [comment]
Same as https://github.com/cms-sw/cmssw/pull/48685#issuecomment-3161484464

**Comment by jfernan2** [comment]
+1

**Comment by missirol** [comment]
Just FYI, with the latest version of this PR (well, of its verbatim 15_0_X backport) there is still a small discrepancy in HLT outputs (see [CMSHLT-3534 (comment)](https://its.cern.ch/jira/browse/CMSHLT-3534?focusedId=7059989&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-7059989)), and I don't know if it is expected.

It might be better to let Tracking POG comment, before this PR is merged.

**Comment by mandrenguyen** [comment]
hold 
Let's wait for tracking POG to comment as requested [here](https://github.com/cms-sw/cmssw/pull/48684#issuecomment-3163239554)

**Comment by mmasciov** [comment]
Thanks for holding. The last commit (here and in the backport PR #48685) should finally address also the last edge case.
For information: this was one case of a single-strip edge cluster. Running over the full test sample by @missirol, I now see zero difference in track and hit/track multiplicities.

**Comment by mmasciov** [comment]
please test

**Comment by mandrenguyen** [comment]
unhold

**Comment by missirol** [comment]
> Thanks for holding. The last commit (here and in the backport PR #48685) should finally address also the last edge case. For information: this was one case of a single-strip edge cluster. Running over the full test sample by @missirol, I now see zero difference in track and hit/track multiplicities.

Just for the record, I can confirm that I see the same ([CMSHLT-3534 (comment)](https://its.cern.ch/jira/browse/CMSHLT-3534?focusedId=7061429&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-7061429)), so the present version of this PR looks okay for what I can check. Thanks, @mmasciov !

**Comment by mandrenguyen** [comment]
+1
